### ISSUE TITLE
SSL server endpoints don't provide a way to verify client certificates

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6757]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ElementalAlchemist reported|
|-|-|
|Trac ID|trac#6757|
|Type|enhancement|
|Created|2013-09-20 20:15:39Z|
|Branch|https://github.com/twisted/twisted/tree/tls-serverFromString-retrieveCertOnly-6757-4|

I've searched and searched, but I can't find a way to set up verification of client certificates from a server through the endpoint interface (using serverFromString rather than calling endpoints directly as recommended).
It'd be nice to be able to accomplish [this](http://twistedmatrix.com/documents/13.1.0/core/howto/ssl.html#auto7) through the endpoint interface.  (Bonus points if I can also make client certs optional.)

I couldn't find any documentation that stated this was possible nor any tickets that said it needed to be (and if such a ticket exists already, sorry about the duplicate).  I took a (very brief) look at the code as well and came up with nothing.

Attachments:

* [6757-proposed1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/733bbfacdc8f3aa32d84a504e71c0cc9225f3bda.diff) (6759 bytes) - added by ElementalAlchemist on 2015-05-03 19:21:42Z - Code proposal for first patch
* [6757-trunk-review1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/488d304f2d8a3dc64be2ca889e1c74e37d4a61d6.diff) (11476 bytes) - added by ElementalAlchemist on 2015-05-09 20:49:49Z - First review copy
* [6757-trunk-review1-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/59aecdf4dffc2cd881d566ad9f990ed2387519f4.diff) (11860 bytes) - added by ElementalAlchemist on 2015-05-09 23:45:02Z - First review copy with topfile
* [6757-trunk-review1-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/7ca7ecfb7c99d0532a83937139b4cef2f60721a2.diff) (12813 bytes) - added by ElementalAlchemist on 2015-05-13 13:49:34Z - 
* [6757-trunk-review1-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/18d63f8f6b1dd3b44c86ab07b8fd9b19268ba4eb.diff) (13890 bytes) - added by ElementalAlchemist on 2015-05-14 00:17:28Z - Third attempt at the first review, now with more docstring
* [6757-trunk-review2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/4cac641b52a9f9ff6858fe1e54abd9c75602bf71.diff) (18414 bytes) - added by ElementalAlchemist on 2015-08-17 04:05:32Z - 
* [6757-trunk-review2-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/df7a41227bdb02d8eb251753fcf336ae9c7a32f3.diff) (18408 bytes) - added by ElementalAlchemist on 2015-09-08 23:40:31Z - 
* [6757-branch3-review3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/bb0435a405eec2d4e8f357e54abcf645ae18e868.diff) (3796 bytes) - added by ElementalAlchemist on 2015-11-01 21:49:28Z - 
* [6757-branch3-review4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/5cc85408e7e6966b93ff50071a13a7192c4106d7.diff) (13220 bytes) - added by ElementalAlchemist on 2015-11-05 02:39:31Z - 
* [6757-branch3-review5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/63fc9f91dbb6eba119efb38b6fe5c37ad6d8293f.diff) (2898 bytes) - added by ElementalAlchemist on 2015-11-06 02:39:17Z - 
* [6757-branch3-review6.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/85687a77dabc216a21c142caf93e8ad863b520f6.diff) (3706 bytes) - added by ElementalAlchemist on 2015-11-13 04:15:43Z - 
* [6757-trunk-review7.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/d8743ba8dde86984226cda33a7fd2bf6cb7e6c5f.diff) (6861 bytes) - added by ElementalAlchemist on 2015-11-29 02:57:27Z - 
* [6757-trunk-review7-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/e18b5566c65e321de578704f375c2334cd9c8567.diff) (7995 bytes) - added by ElementalAlchemist on 2015-11-29 06:31:26Z - Replacement for the previous patch that includes documentation updates
* [6757-branch4-review7.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/b712a6e6e588c6422e14413eb08aa871b2f24927.diff) (7748 bytes) - added by ElementalAlchemist on 2015-11-29 19:02:37Z - 
* [6757-branch4-review8.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46a/46a04acaf0998e1be38bce1655950eaaf758cb1d/8dc27b84bb04384d87cbe4f3d75383666f97c418.diff) (7686 bytes) - added by ElementalAlchemist on 2015-12-14 06:05:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6757 6757
type__enhancement enhancement
reporter__ElementalAlchemist ElementalAlchemist
priority__normal normal
milestone__ 
branch__branches_tls_serverFromString_retrieveCertOnly_6757_4 branches/tls-serverFromString-retrieveCertOnly-6757-4
branch_author__adiroiban__glyph adiroiban, glyph
status__new new
resolution__None None
component__core core
keywords__None None
time__1379708139000000 1379708139000000
changetime__1450189326434147 1450189326434147
version__None None
owner__ElementalAlchemist ElementalAlchemist
cc__lukasaoz@...
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ElementalAlchemist commented|
|-|-|

This patch isn't done yet, but I wanted to see if I could get some preliminary feedback before I get too invested.
The code is done except for my question below and any feedback which may help to improve it.

TODO:
* Write tests
* Update documentation for serverFromString

QUESTIONS:
* There's not currently a good way to represent boolean values for serverFromString parameters (since it's all one string). Should we develop one for this?

Proposal coming in the first attachment. I'll start doing some test updates later today, and I'll mark this for review when that's done and a patch with tests is uploaded. I'm saving the documentation until the end of the process so that I at least don't have to write the documentation three times in case of any major revamps.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I'd suggest writing the user-facing documentation and examples first (or at least next) to demonstrate that the proposed interface is actually simple and provides the desired features.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ElementalAlchemist removed owner|
|-|-|

I have a couple test-related things to figure out (namely the todos for the tests marked todo), but in the meantime this is probably OK for an initial review. This includes a bit of an update to the doc I linked. I didn't update any of the other docs because I didn't see any other docs that talk about that, and not everything I just did is there because it didn't all seem to make sense in the context of that doc.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ElementalAlchemist commented|
|-|-|

Added a version with a topfile

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Hi,

Thanks for the patch.

At this time I am only checking the docstrings without looking at the actual implementation.

ssl_clientauth_server_endpoints.py is missing from the patch.

I fail to understand why you would need this functionality and what is the problem which this patch tries to solve.

OK you want to write a set of python lines as a single serialized string, but why ? :)

The documentation was updated and the code referenced in the ticket description is no longer part of Twisted documentation.

Latest docs: http://twistedmatrix.com/documents/current/core/howto/ssl.html#client-authentication

-------------

I find it hard to understand this docstring:

```
@param retrieveCertOnly: When C{verify} and C{caCerts} are not 
   specified and this parameter is L{True}, sets the certificate 
   options such that peer certificates are retrieved. This 
   parameter is L{False} by default. 
```

First I find that the precondition is complicated and in the code I can see that it overwritten by `verify`

Peer certificates are retrieved... and then what? :)

How can I specify a callback which will get a copy of the certificate?

```
def _retrieveCallback(conn, cert, errno, depth, preverify_ok): 
```

needs a docstring as otherwise I have no idea why we need it.

in twisted/internet/endpoints.py `requireCert` is not documented

I could not find the docstring which talks about setting 'yes/no' values for boolean flags in endpoint string serialization.

------------

I am leaving this in the review queue so that other developers can check it.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ElementalAlchemist commented|
|-|-|

Dangit, svn.
I'll upload something now to add that file, and I'll start addressing your other comments later today.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ElementalAlchemist commented|
|-|-|

> I fail to understand why you would need this functionality and what is the problem which this patch tries to solve.
The problem is that when I call a user's `transport.getPeerCertificate()`, it always returns `None`.
Before moving to endpoints (which are "the future"), we had the ability to retrieve client certificates by setting some certificate options parameters.
My application is an IRC server, so it's useful for us to be able to retrieve client certificates (ideally without the requirement that the client cert must be issued by a CA, since most client certs on IRC are self-signed). (If by bypassing the CA check I'm also bypassing something else super important, I'd love to know that--but we should also try to work together to figure out how to resolve it.) While I was there, I figured I'd add the `verifyCACerts` parameter that's already available in certificate options, since it's likely to be more useful to other developers.

> OK you want to write a set of python lines as a single serialized string, but why ? :)
I don't know what you mean.

>The documentation was updated and the code referenced in the ticket description is no longer part of Twisted documentation.
>Latest docs: ​http://twistedmatrix.com/documents/current/core/howto/ssl.html#client-authentication
I compared with the newest docs (from the "Go to the latest version of this document" link at the top), and the patch is against whatever was in trunk at the time. That was the newest doc when I first wrote this ticket, and it doesn't look like it's been fixed in that time.

> I find it hard to understand this docstring:
> ...
> First I find that the precondition is complicated and in the code I can see that it overwritten by verify
This is fair. I made an attempt at clarifying the docstring (see next patch).

> Peer certificates are retrieved... and then what? :)
And then the application can look at/use it? Its lack of availability is the issue currently--as stated above, without it `transport.getPeerCertificate()` always returns `None`. This allows `transport.getPeerCertificate()` to return something useful. (I'm talking about [this guy](http://twistedmatrix.com/documents/current/api/twisted.protocols.tls.TLSMemoryBIOProtocol.html#getPeerCertificate).)

> How can I specify a callback which will get a copy of the certificate?
I don't know what you mean here, either.
If you want to pass a custom callback... I don't know exactly how that would work, given that `serverFromString` takes a string. If we really think that'd be useful, I'm open to ideas.

> [_retrieveCallback] needs a docstring as otherwise I have no idea why we need it.
Added.

> in twisted/internet/endpoints.py requireCert is not documented
Whoops! I added that parameter after I did the docstrings the first time and forgot to reupdate them. Added.

> I could not find the docstring which talks about setting 'yes/no' values for boolean flags in endpoint string serialization.
I've added more docstrings relating to this as well, but I'd love to know if I can explain more about them.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Thanks for the update.

I think that the example should be more verbose and serve
as narrative documentation.

For example is should highlight that `requireCert=yes` is the highlighted part
of this example and that it can get yes/no values.

Without your comments from the ticket I am still confused :) The example don't make use of transport.getPeerCertificate()
and as far as I understood this ticket is all about making transport.getPeerCertificate() return the remote certificate.

So I think that the example should also highlight the use of getPeerCertificate().

I don't see the getPeerCertificate() call in the tests so I still find it hard to understand what is the problem with current code from trunk... again, just be reading the code and ignoring your comments from this ticket :)

--------

Please not that I know little about endpoint string serialization and I just try to raise some general questions to make the life easier for future reviewers.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

I think that we need a separate ticket for `verifyCACerts` as this change does not match with the ticket title or the changelog entry.

Also, this ticket is about validating client certificate or retrieving/making it available ?

From your commend I understand that the changes are about making the certificate available in getPeerCertificate() so that custom code can implement custom checks.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ElementalAlchemist commented|
|-|-|

The documentation I updated is about using endpoints to set up connections. A side note about `TLSMemoryBIOProtocol.getPeerCertificate` would probably be out of place. Would you prefer I write a new doc? Where should it go?

I added the part about `verifyCACerts` since it seemed like what was intended to be the right way to go and was trivial to add--most of the patch is about getting `requireCerts` going. If we're sure it's worth separating out, `verifyCACerts` can go in its own ticket.

This ticket is primarily about retrieving the certificate. The TLS code requires some setup to be able to retrieve client certificates, and this setup isn't done at all by default, nor was the ability to do so exposed through `serverFromString`. This is about exposing a way to set up the TLS protocol object so that later, when we're communicating with the client, we can then look at and use the client cert.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ElementalAlchemist removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

(In [45102]) Branching to tls-serverFromString-retrieveCertOnly-6757.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

I have created a new branch, applied the patch and sent it to the buildbot.

You should either fix the .todo tests before merge or remove it from the patch and work on in a separate ticket.


You can check the results... but with .todo tests which are skipped I find it hard to see if the patch is valid on all supported platforms.

Let me know if you need help with updating the tests.

----------------

I think that you can also use trunk to set up connections using endpoints.

In both the new example and the test file I was expecting to see that some call is done to retrive the certificate (ex transport.getPeerCertificate() is called) and for the case when the flag is not set, it will return None while for the case when it set it returns the expected certificate.

I still consider my notes from comment:8 valid,

I am leaving this in review so that another developer could check your changes and maybe your patch will be accepted.


Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|ElementalAlchemist commented|
|-|-|

> Let me know if you need help with updating the tests.

The .todo tests are todo because we need more test certificates (specifically a test CA cert and test server/client certs generated by that CA). I don't know if there's a special process for adding certs or something. (Also I'll need to figure out how to do the certs correctly from the CA.)

> In both the new example and the test file I was expecting to see that some call is done to retrive the certificate (ex transport.getPeerCertificate() is called) and for the case when the flag is not set, it will return None while for the case when it set it returns the expected certificate.

I still think it'd be out of place in the doc I updated and in any of the other docs I found. Writing an entire new doc on handling SSL connections on a server may or may not be a bit overkill.
I could, however, reasonably write a test for that. Is there an example test of a client and server communicating or otherwise a good way to simulate that in the test suite? This would also depend on that first thing I said where certs for the server/client are needed (or maybe we could use the existing ones for this one; I forget what's there at the moment).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

(In [45471]) Branching to tls-serverFromString-retrieveCertOnly-6757-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

The branch needs merging forward so I am going to do that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to ElementalAlchemist|
|-|-|

Thanks again for your work, ElementalAlchemist.  This is a big hole in our TLS story and I'm glad you're fixing it.

Unfortunately there are still some issues, and once again they're not really all your fault :).

    1. I think I addressed the conflicts incorrectly when merging forward, so apparently they weren't so trivial.  The buildbots are failing, so please submit another patch against trunk.
    2. Please don't add any new `.todo`'d tests.  Long term, we are trying to get rid of all the `.todo`'d tests on trunk, and eventually remove the feature from Trial entirely.  `.todo`s are a record of work that needs to be done, and records of work that needs to be done belong in an issue tracker, not a test suite.  Just file a new ticket and attach the test to it.
    3. How is `retrieveCertOnly` supposed to be used in a real application? This seems like a security land-mine designed to trap the unwary.  This needs to be documented more clearly in the narrative documentation.
        1. If this is in fact a good idea, rather than documenting "`this parameter only has an effect when C{verify} and C{caCerts} are not specified`", you should make the parameter mutually exclusive with `verify` and `caCerts` as well.
        2. It also shouldn't be a public variable; the API surface for `CertificateOptions` is already ridiculously broad.  Just call it `_retrieveCertOnly`.
    4. `_getCertificatesFromFile` is a nice refactoring, but it needs a docstring to help out future maintainers, as I'm somewhat surprised `twistedchecker` did not point out...
    5. I think symmetry with the client endpoint would be a good idea.  On the client, the parameter is simply `caCertsDir`.  I think we should evolve them in lockstep, adding a `caCertsDir` to the server, then adding `caCertsFile` to both if we want to add multi-certificate files at the same time.  Random differences like this are a good way to confuse users.  Similarly, the way that `caCertsDir` implies `verify` on the client, it ought to imply `requireCert` on the server.

My suggestion for moving forward with this would be to submit one ticket for the straightforward refactorings (`_getCertificatesFromFile`, `_valueToBool`) and then do the certificate-trust stuff in a future branch.

There should also be a bit of narrative documentation on the server side saying what the expected security properties are of these various arguments.  Supplying such information is crucial for users to actually get the security they want from their endpoint configuration.

As usual, sorry for the long review cycles and the feedback that didn't necessarily get taken into account in previous reviews.  On the bright side, some of this is simply rephrasing adi's point from comment 8, so not all of it is new :).  Thanks again, and I hope that we'll see a resubmission from you soon!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|ElementalAlchemist commented|
|-|-|

1. Can do.
2. I'd intended for the `.todo` tests not to be `.todo` before the patch was finally complete. They're marked as that for now so we notice that they need to be completed before the patch is merged into trunk. Is there some sort of process for adding test certs, or should I just make some?
3. I'll make a doc just about `serverFromString` (or maybe about both `*FromString`s or whatever seems to make sense when I get back into the documentation layout). For my specific case: as an IRC server, I see a lot of client certificates. Many of them aren't signed. The ones that are will generally be signed somewhere like CACert, which, while a legitimate CA, isn't really in a lot of master trust files. In these cases, it'd still be nice to be able to get cert information from the client. (I don't know much yet about what other verification does or doesn't go on behind the scenes (e.g. is there at least a challenge to see if the client has the matching key?), so you might be able to convince me this is a bad idea as it is, but I'd like to see if we can work through all that issue.)
  1. Can do.
  2. OK.
4. Can include in patch update.
5. Makes sense; inconsistent APIs are the worst. I can look into clientFromString and see what I can do with it.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|ElementalAlchemist removed owner|
|-|-|

Added review2. In addition to the issues mentioned in the review, I also found a number of screw-ups of my own, so hopefully that's all better now.

1. I reapplied the changes on the current trunk. The patch should merge fine until the trunk changes significantly again.
2. I managed to un-todo all my todo'd tests (yay!). It turns out we didn't actually need any new test certs, so the tests could be finished as they were. (If, however, we can test a full connection and I can figure out how to do that in trial, generating some test certs for more comprehensive tests might be beneficial.)
3. I found the endpoints doc and added some documentation relating to the new `serverFromString` parameters. It mentions the security implications, though I can go more in-depth if it's necessary or otherwise important.
4. Added a docstring.
5. I switched the server parameter to `caCertsDir`. Maybe if we want `caCertsFile` it can be added later. As for the last part of this point, not implying `requireCert` is the reason I didn't just implement `trustRoot`. Unless you mean "if `caCertsDir` is specified, `requireCert` should be on by default unless turned off"?

As for the other points:
- I can separate the `_getCertificatesFromFile` refactor if you think it's necessary. `_valueToBool` was introduced for the new arguments in this ticket, so if we introduce it separately, it probably won't be called from anywhere until this is merged.
- The bit about documentation I believe I discussed in point 3, and, as mentioned, I'll add more detail if we think it's important.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl commented|
|-|-|

Not a review, but just want to note that serverFromString works on ```str``` now, not bytes.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|ElementalAlchemist commented|
|-|-|

Updated docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

(In [45688]) Branching to tls-serverFromString-retrieveCertOnly-6757-3.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph commented|
|-|-|

Patch applied, builders spinning.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_25"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Hi,

Sorry for the delay.

First, I would like to review the documentation, so that we know what is the scope and purpose of this ticket matched the ticket title and description.

`verify` option, which will exclude `retrieveCertOnly` is not documented in the endpoints.rst

I think that it should be documetented and also talk about / document the mutual exclusive arguments.

------

I think that the endpoints.rst should talk about `retrieveCertOnly` not `retrieveCerts`.

-------

`functions when ``caCertsDir`` is not used.` I don't understand these words.


------

In the  `requireCert` documentation, what is a `valid certificate` ?

---------

How do `extraCertChain` interact with the `caCertsDir` option?

Can we have a single option? .... for example auto-detect if the path is a directory and load 
all files from that directory.

I would prefer to keep the endpoints options to a minimum.

---------

Why not set up the context so that the certificates are alwasy retrieved (alwasy set SSL.VERIFY_PEER)... then we can just return True if retrieveCertOnly is True... but later they are still available under all circumstances.

I would prefer to have a single callback for ctx.set_verify and include all the login inside that callback, rather than in the context creation.


We would no longer need the `retrieveCertsOnly` option and just use `requireCert` ... or rename it to
`requireValidClientCert`... or something more descriptive :)


-------


Please consider the example file as part of the  documentation and don't worry if the comments are to verbose.
It is fine (and I would say recommend) for the examples to have comments which talk about each call and why and how it is made.


Please check my review and let me know what do you think.

Once the documentation is agrees I can continue to review the code.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|ElementalAlchemist removed owner|
|-|-|

> verify option, which will exclude retrieveCertOnly is not documented in the endpoints.rst
> 
> I think that it should be documetented and also talk about / document the mutual exclusive arguments.

> I think that the endpoints.rst should talk about retrieveCertOnly not retrieveCerts.
Neither of these are related to endpoints; both are discussing changes to `OpenSSLCertificateOptions`. `OpenSSLCertificateOptions` isn't really described in depth in narrative documentation, nor do I think it necessarily needs more than the API docs. It's more of an implementation detail for SSL endpoints, and it's not described in the narrative documentation for endpoints because users of endpoints don't directly interact with it. The parameters described with the names documented are the names of the parameters to `serverFromString`, which is the thing with which users **do** directly interact.

> functions when caCertsDir is not used. I don't understand these words.
I see what you mean. I've clarified this. (Upload coming right after this post.)

> In the requireCert documentation, what is a valid certificate ?
Clarified this as well.

> How do extraCertChain interact with the caCertsDir option?
These don't do even remotely the same thing.

``extraCertChain`` allows you to specify extra certificate information for **your** certificate which you're passing with the ``certKey`` parameter to have a server-side certificate.

``caCertsDir`` is a directory of CA certificates that you trust and can validate against for certificates you receive.

I welcome feedback on naming for ``caCertsDir``, but I don't think we can just make them one parameter since they fundamentally do very different things.

> Why not set up the context so that the certificates are alwasy retrieved (alwasy set SSL.VERIFY_PEER)... then we can just return True if retrieveCertOnly is True... but later they are still available under all circumstances.
I really like this idea. ``OpenSSLCertificateOptions`` is (I imagine) used a lot in a lot of places, so I'd love feedback from someone who knows better whether it's safe to just do this. I've always found it unintuitive that there's just a function exposed that always returns ``None`` unless you do something... and this would fix that.

> Please consider the example file as part of the documentation and don't worry if the comments are to verbose. It is fine (and I would say recommend) for the examples to have comments which talk about each call and why and how it is made.
I added some comments. I also fixed how out-of-date it was.

-----------------

I'll submit the documentation updates I made now. I look forward to the next review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_27"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Many thanks for the latest patch! I have applied it on the branch and send it to buildbots.

I am still only reviewing the documentation.

------

Sorry about the OpenSSLCertificateOptions confusion and thanks for the explanation.

-----

Thanks for the `extraCertChain` explanation.

Maybe instead of `your certificate` we can update to docs to say `server's certificate`.

------

If you also think that the server endpoint should always try to retrieve the client certs, then I think you should go for it.

Feel free to remove the `retrieveCerts` option and always set up the context to get the cert. less is more :)

-------

Do we also need the `requireCert` options? I was expecting that once `caCertsDir` is set all certificats will be checked against those CAs.

Why would you want to set caCertsDir without requireCert ?

Again, less is more :)

------

caCertsDir. Can we rename it to clientCACerts or something to hint that this is used only for the client certs? 

-----


What do you say if we rename `caCertsDirs` to just `caCerts` or `caCertsPath`. In this way we can reuse it later for the case when we want to have a single PEM file with multiple CAs.

For now, we can document that it is a directory... but later maybe it can be extend to also be a file.

-------

For twisted/internet/test/test_endpoints.py can we move all SSL related tests into a commmon test suite and then have a single skip for the whole test suite class?

----------

If I run the example as

```
$ python docs/core/examples/ssl_clientauth_server_endpoints.py
```

I get 

```
2015-11-03 13:50:45+0200 [-] IOError: [Errno 2] No such file or directory: '/home/user/twisted/server.pem'
}}

maybe we can use something like this

{{{
diff --git docs/core/examples/ssl_clientauth_server_endpoints.py docs/core/examples/ssl_clientauth_server_endpoints.py
index e43b847..17688e0 100644
--- docs/core/examples/ssl_clientauth_server_endpoints.py
+++ docs/core/examples/ssl_clientauth_server_endpoints.py
@@ -6,6 +6,7 @@ import sys
 
 from twisted.internet import defer, endpoints, protocol, task
 from twisted.python import log
+from twisted.python.filepath import FilePath
 
 import echoserv
 
@@ -24,7 +25,9 @@ def main(reactor):
     #   requireCert=yes: This makes the socket reject client connections that
     #     do not provide a certificate that passes validation using the CA
     #     certs in caCertsDir.
-    descriptor = "ssl:8000:caCertsDir=.:requireCert=yes"
+    descriptor = (
+        "ssl:8000:privateKey=%s:caCertsDir=.:requireCert=yes" % (
+            FilePath(__file__).path,))
     # Pass the reactor and descriptor to serverFromString so we can have an
     # endpoint
     endpoint = endpoints.serverFromString(reactor, descriptor)
```

This still fails, but the idea is that we should have the examples as self contained as possbile.

If the server.pem is not ok, maybe we can update the comment to make it clear that users will have to brink their own cert file and put it at a specific path

--------

Please check my comment and submit a new patch with updated documentation... if your are confident in the new documentation feel free to also update the tests and the code.


Thanks again for your work!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|ElementalAlchemist removed owner|
|-|-|

> Thanks for the extraCertChain explanation.
> Maybe instead of your certificate we can update to docs to say server's certificate.
Done.

> Feel free to remove the retrieveCerts option and always set up the context to get the cert. less is more :)
Done.

> Do we also need the requireCert options? I was expecting that once caCertsDir is set all certificats will be checked against those CAs.
> Why would you want to set caCertsDir without requireCert ?
`caCertsDir` makes it so that the client certs are checked against some CA certs. `requireCert` makes it so that a valid cert (according to those CA certs) is required in order for the connection to complete; without a valid cert, the connection is rejected before we see the connection (at least on the app end; not sure how much of it Twisted sees). Without `requireCert`, clients are still able to connect if they provide no certificate.

`requireCert` being enabled automatically in that case is the reason I did not implement `trustRoot` instead.

> caCertsDir. Can we rename it to clientCACerts or something to hint that this is used only for the client certs?

> What do you say if we rename caCertsDirs to just caCerts or caCertsPath. In this way we can reuse it later for the case when we want to have a single PEM file with multiple CAs.
Renamed to `clientCACertsPath`.

> For twisted/internet/test/test_endpoints.py can we move all SSL related tests into a commmon test suite and then have a single skip for the whole test suite class?
I don't know enough about trial to finagle this one in. Do I need to do anything to hook the test suites in, or is it just as simple as creating another class with some `test_` methods?

> If I run the example as...
Updated this not to care where it's run from (though you still can't copy the script elsewhere and run it without also moving some other files).

Upload coming right after this post submission.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_29"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Thanks for the latest changes

-------

Maybe we should rename the example as tls_clientauth_server_endpoints.py ... I think that SSL is no longer used by default.

The example needs some instructions about how to set it up and how to use it.

When posible we should instruct users to create their own crytpo material rather than instructing them to re-use some RSA private keys with an expired cert.

I have extended the example to give some instructions about how to use and how to set it up.

It would be nice if the server will also show connected / disconencted clients together with their disconnection reason

When I set up requireCert=no and the client will send a cert, the connection will fail.

```
2015-11-05 10:26:27+0200 [-] Server started as ssl:8000:privateKey=server-key.pem:certKey=server-cert.pem:clientCACertsPath=ca-certs:requireCert=no
2015-11-05 10:26:27+0200 [-] Connect to it from another terminal.
2015-11-05 10:26:29+0200 [-] IPv4Address(TCP, '127.0.0.1', 46610) Connection made
2015-11-05 10:26:29+0200 [-] IPv4Address(TCP, '127.0.0.1', 46610) Connection lost: [('SSL routines', 'SSL3_GET_CLIENT_CERTIFICATE', 'no certificate returned')]

```


I need that example in order to do a manual functional test and check interoperability with other SSL implementations.

We should also add unit tests for the new code from _sslverify... but for now I only want to focus on the docs and UI/UX.

---------

I moved all ssl related tests into a common test case so that we no longer have to define a new skip for each new TLS related test.

------

I don't understand why trustRoot is not ok.

Please expand the docs/core/howto/ssl.rst docs to explain why we would need caCertsDir=some-path and requireCert=no

------

Please check my latest changes on this branch.
I have just moved the test + updated the example file to explain how to use it ... or how I used it :)

I have also updated the example to use a server protocol which gives us more info about the connections

--------

endpoints.rst still talk about caCertsDir maybe endpoints.rst should not list all available options but rather just refer to the API docs.

I don't see much value in duplicating that info ... as in this part, the narrative docs is just describing (again) each endpoint option.

------

I still don't understand how this feature should work.
It needs better documentation to act as specification.



Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_30"></a>|ElementalAlchemist removed owner|
|-|-|

> Maybe we should rename the example as tls_clientauth_server_endpoints.py ... I think that SSL is no longer used by default.
Maybe we should rename the endpoint type to "tls" since SSL is no longer used by default.

I think there's value in this case of using the consistent naming conventions.

> I have extended the example to give some instructions about how to use and how to set it up.
I like it.

> It would be nice if the server will also show connected / disconencted clients together with their disconnection reason
I'd copied one of the existing examples and modified it. I thought these just needed to be included in documentation. I think your edits are good.

> When I set up requireCert=no and the client will send a cert, the connection will fail.
Was the cert signed by one of the CA certs?

> We should also add unit tests for the new code from _sslverify... but for now I only want to focus on the docs and UI/UX.
I'd love to have tests where we create an endpoint and then try to connect various clients to it if I can figure out how to do that.

Noted about needing more tests. Now that reviews are getting done, maybe I won't forget between reviews. :P

> I don't understand why trustRoot is not ok.
Using IRC as an example: We may want to use a certificate for oper authentication without requiring all other users to also have a signed certificate.

> endpoints.rst still talk about caCertsDir maybe endpoints.rst should not list all available options but rather just refer to the API docs.
> 
> I don't see much value in duplicating that info ... as in this part, the narrative docs is just describing (again) each endpoint option.
How easy is it to point to the API docs of a class in a private module (_sslverify)? Otherwise, I fixed the name.

> I still don't understand how this feature should work. It needs better documentation to act as specification.
The point was for the functionality we originally had to happen. I figured I'd add verification while I was there, since it wasn't a lot extra and it sounded really useful to someone. Otherwise, people like me are told Endpoints Are The Future™ so we switch to them and then lose a whole ton of features (like being able to tell our SSL sockets to verify certificates).

--------
Patch incoming for the doc parameter name fix.

--------
Dear SpamBayes,

I can certify for sure that this comment is not spam. Please stop that.

With love,

ElementalAlchemist

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_31"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

OK. Let's keep file name as ssl.

I have done a small refactoring in the _sslverify code as I feel that code can be simplified... and I was trying to see why the client certs are not retrieved... forget about the handshake delay :(

-------


I tried to test the code as described in the examples ... now it works.

I did something wrong.

I have updated the example to also show the peer cert.

------

I can help with writing the unit tests... but first lets make sure that the design and specification of the new features are ok.

We will use the documentation to implement the tests :)

-------------


For the IRC example, why not use just this ?

```
descriptor = ('ssl:8000:privateKey=server.pem:')
```

The peer certificate is alwasy retrieved.

Then in your code you can get and check the cert

```

def irc_AUTH(self):
    peerCertificate = self.transport.getPeerCeritifcate()
    self._validateCertificate(peerCertificate)

```

Can you please expand this use case in the narrative documentation?

I agree that we should set up the context so that the peer cert is alwasy retrieved. I fully understand this requirements and I can image how it can be usefull.


--------

Is very easy to point to API from narrative docs. Use the

```

:api:`twisted.internet.some.Class.method`

```  

If you need help just put some code and methion to were you want to link to an API and I can update it 

--------


Please check my comments. Sorry for all the ping-pong. Feel free to get in touch over irc, I am in the GMT timezone.

Once the documentation and user experience is defined we can look at implementing it.

Many thanks for your feedback!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_32"></a>|ElementalAlchemist removed owner|
|-|-|

I adjusted the narrative docs to point at the API and to explain a little bit more of when `requireCert` is useful. If my explanation is unclear, please let me know. Also, definitely let me know if the API link won't work (it probably won't since private APIs are hidden by default) and what to do about it (or if restoring the old explanations is OK).

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_33"></a>|@Lukasa commented|
|-|-|

Adi asked me to leap in to help reviewing, so here I am. =)

----

First things first, the example filename. This is a little thing, but I added a couple of examples recently, and they began `tls_`, not `ssl_`. Might be better to be consistent here?

It seems a little bit weird that we explicitly call out that this is a *client* CA cert path. I suppose I can see the value in it, but it's a little odd.

I don't have any particular objections to the style, however I share some of adiroiban's concerns about overloading the descriptors. I think I'd be less worried if they were keyword arguments to a function. However, I'm not a twisted expert so I'm substantially less sure than Adi is.

From a TLS layer perspective this design makes sense. I'm also in favour of allowing the option to validate certificates without forcing them to be presented.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_34"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Hi, 

Thanks lukasa for the review!

ssl_ is there to be in sync with the other ssl_endpoints examples.. and the TLS endpoint descriptions starts with "ssl:" :)

---------

ElementalAlchemist, sorry but for me the current docs are still not clear... but I think I understood the need for this.

So the requirements for the IRC use case should be:

* You have an IRC server over TLS
* You want to allow any users to connect to it, even if they don't have a SSL cert.
* Any non-admin commands can be requested for clients not using SSL client certs.
* All admin commands should be issued by clients connected using SSL with a client cert signed by a specific CA.
* You want to allow admins with SSL certs signed by multiple - non related - CAs

--------

idnar commented on IRC that at least for HTTPS, if you always set SSL.VERIFY_PEER the web browser will always pop a window/dialog if you have a client cert imported in the browser... and there is no configuration option to restrict to usage of a client Cert to  a list of websites.

I think that this is a client / browser issue, nevertheless we should take this into consideration.

So the description API should also accommodate this requirement:

* In case you never need to get the client cert, the connection should not ask for it.


----------

I am a bit worried about putting too many options in the description... and putting options which are not generic.

From the ticket description, I see that the main issues is that the endpoint does not provide a way to verify client certs.

From this, I understand that it should provide a method for generic checks.

What do you say about:

```

endpoint = endpoints.serverFromString(
    reactor, 'ssl:port=1234:getClientCertificate=yes')


class IRCCommands:

   def irc_admin(self, optinos):
      clientCert = self.transport.getPeerCertificate()
      if not clientCert:
          return self.quit('You need a cert.')
      if not self._validateClientCert(clientCert):
          returl self.quit('Invalid client cert')
      do_admin_stuff()
```

Later we can extend it to something like this, so that you can do validateClientCert in the SSL handshake

```

endpoint = endpoints.serverFromString(
    reactor, 'ssl:port=1234:getClientCertificate=yes')
ITLSEndpoing(endpoint).setPeerValidation(_validateClientCert)

```

_validateClientCert can load CA from local disk or from other locations... can also load CRL from disk/http with refresh or from CRL Distribution Points... it can also ask for OCSP..etc 


I can imagine the use case of an long running HTTPS server in which I want to re-load the allowed CAs + do CRL via CDP + do OCSP checks at runtime.

A static list of CA certs stored on local filesystem will not help to much in this use case and we will end up with adding yet another config option to suit this need.

I think that configuring an endpoint from a string description to ask for client cert is fine.
Configuring it with all kind of complicated logic, is not and this should go in python code. 

-----------


What do you think?

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_35"></a>|ElementalAlchemist removed owner|
|-|-|

I think if we're going to do all that, we can punt most of it until later? My primary goal here was to be able to retrieve client certs at all. If we want to deliberate on it for a while, I can put off certificate verification until a later time. (If we want to tackle it now, I can rework `_parseSSL` and `SSL4ServerEndpoint` to try to support this reasonably.)

For if we do end up doing it here: If we're worried about long-running processes getting updated CA certs, rebinding ports with endpoints is reasonably trivial, so in the case where we don't do anything extra about it, it's still reasonably easy to make it work. (I understand if we want to make it easier than that, but in the worst case, it's not hard.)

I'm submitting a new patch based on trunk which takes us back to basics and just does certificate retrieval. If we want to do CA verification of certificates, let's discuss it for a bit (possibly on IRC where we can have such a discussion at a reasonable pace without a bunch of tag finagling here, and we can paste the chat log in here after).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_36"></a>|@adiroiban commented|
|-|-|

Many thanks for your patch!

I am very happy to see progress on this :)

It looks good. I will look into details and applying it a bit later.

Feel free to open a ticket and discuss various ways in which the certs can be auto-checked...but I think that it would end with some "framework" and an ICertValidationPolicy interface.. and things might get complicated.

So maybe we can start with a simple API like this and people can implement their own policies and we can provide some standard policies as functions or callables

```
endpoint = endpoints.serverFromString(
    reactor, 'ssl:port=1234:retrieveCertificate=yes')
ITLSEndpoing(endpoint).setPeerValidation(_validateClientCert)
```

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_37"></a>|ElementalAlchemist commented|
|-|-|

Created [#8134](https://github.com/twisted/twisted/issues/8134) to discuss/eventually solve the certificate verification stuff.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_38"></a>|@adiroiban commented|
|-|-|

(In [46366]) Branching to tls-serverFromString-retrieveCertOnly-6757-4.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_39"></a>|ElementalAlchemist commented|
|-|-|

New patch incoming; includes all the stuff we did before that I forgot about that was mentioned on IRC (topfile, etc.) and increases test coverage.

The following is still needed:
* Move all the SSL tests to a single test case
* Comprehensive test of the server actually getting a connection from the client and getting a certificate from `getPeerCertificate`

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_40"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Hi,

Many thanks for the latest patch.
I think that we are close to having this merge :) Hooray!

I have applied your patch.

While reviewing it I made a few changes, including updating it to use getClientCertificate everywhere. I have also updated the docstring so that the exception raised for mutually exclusive getClientCertificate arguments is documented in the dedicated section.

------

I have also refactored the verify callback in an attempt to simplify the code... and I don't think that we need to have custom check against self.verify as the decorator should handle it.

Please let me know if they make sense, otherwise feel free to reverse them.

-------

I think that for _valueToBool returning both None and bool is a bad API.
Instead of returning None, I have just raised the error.... as from what I can see you were using None to later raise the same error.
 
Also, I don't understand the need for a default value... as before calling _valueToBool we check for getClientCertificate.

If only yes/no values are documented, why do we have code for supporting true/false? We should either document true/false or remove the code. I prefer to have as few code as possible as it is easier to maintain and test.

We should not just check that any TypeError is raised... but our TypeError :)... I have updated the test.

------

I have move the SSL server tests into a dedicated test case so we don't need to add skip for each tests and we don't need to prefix each test with ssl or with server.

------

I don't see any tests for the changes from _sslverify.
We need unit tests for the mutually exclusive options and functional tests to check that getClientCertificate actually works.

------

You can check the NPN/ALPN tests for some hints about how to implement the functional tests.

Let me know if you need help with writing the missing tests.

Please check my latest 2 commits on this branch.

I think that the patch only needs to missing sslverify tests.

I will also try to find more people to check that docs and high level API is ok but this should not be a blocker.

Looking forward for the next patch.

Many thanks for your contribution!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_41"></a>|ElementalAlchemist removed owner|
|-|-|

`OpenSSLCertificateOptions` is used by non-server SSL connections, too. I changed the parameter/variable name to `getPeerCertificate` so it makes sense for both.
Everything else you did looks good.

I like the idea of the verify callback stuff, but scoping in Python is weird. Are you sure the right `self` will be used? I included this in my patch; feel free to rip it out if you're positive yours will always work.

`_valueToBool` was meant to be generally useful in case it comes up later. I'm fine with making the caller check for defaults; that was a holdover from before when there were a couple different arguments taking yes/no values. I'm fine with your edits except that I'm making part of the error string a parameter so that different `serverFromString` arguments can require yes/no values. Let me know what you think.

I'm working on the overall "make a connection and run `getPeerCertificate()` on the client transport" test. It looks like there's some infrastructure in twisted/tests/test_sslverify.py that can kinda be used, but we'll have to make it use endpoints to test this (which means not doing it there). I'll continue working on that particular test, and it's not in the patch, but I think the rest should be OK. Let me know what you think. Patch incoming immediately following this post.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_42"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Your changes to _valueToBool are great but they need docstring. I tried to add them and I hope that new argument name better express its usage.

-------

If you think that scoping don't work, please write a test for it. 
I have reverted the changes.


I find Python scoping easy to understand. http://stackoverflow.com/a/292502/539264


------

Looking forward for the full tests for sslverify 

Let me know if you need help with writing the test.. you can always ask for help also on #twisted-dev

Many thanks for your contribution.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_43"></a>|ElementalAlchemist commented|
|-|-|

Oops! Yes, thank you for getting the `_valueToBool` docstring. That's right.

I used the wrong term for that. Binding of nested functions is weird in that Python generally does it when the function is called rather than when it's declared (as far as I can tell). Scoping is super easy, but the context of nested functions' scopes is something I still don't get even after all this time with Python.

I'll continue working on the full test and possibly more sslverify tests as well.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_44"></a>|@adiroiban commented|
|-|-|

Thanks! Looking forward for the next patch.

I understand your point of view but in the same time I prefer not to write code in the spirit of FUD (Fear, uncertainty, and doubt) ... and have tests to kill any FUD feeling.

So if you feel that something might not work (now or in the future) please write a test for it. It might require more work, but for the long term I hope it will be better than having a code-base full of untested things written based on FUD ... and in the process of writing the test you will gain a better understand of what is going on there... and people reading the tests should also gain a better understanding of that code.

Regards,
Adi
